repose-aggregator/components/filters/openstack-identity-v3/src/main/scala/org/openrepose/filters/openstackidentityv3/OpenStackIdentityV3Handler.scala

```diff
@@ -251,17 +253,38 @@ class OpenStackIdentityV3Handler(identityConfig: OpenstackIdentityV3Config, iden
   private def containsRequiredEndpoint(endpointsList: List[Endpoint], endpointRequirement: Endpoint) =
     endpointsList exists (endpoint => endpoint.meetsRequirement(endpointRequirement))
 
-  private def writeProjectHeader(projectFromUri: String, roles: List[Role], writeAll: Boolean, filterDirector: FilterDirector) = {
-    val projectsFromRoles: Set[String] = {
-      if (writeAll && roles != null) {
-        (roles.collect { case Role(_, _, Some(projectId), _, _, _) => projectId} :::
-          roles.collect { case Role(_, _, _, Some(raxId), _, _) => raxId}).toSet
-      } else {
-        Set.empty
+  private def writeProjectHeader(defaultProject: Option[String], roles: List[Role], projectFromUri: Option[String],
+                                 writeAll: Boolean, sendQuality: Boolean, headerManager: HeaderManager) {
+    lazy val projectsFromRoles = (roles.collect { case Role(_, _, Some(projectId), _, _, _) => projectId} :::
+      roles.collect { case Role(_, _, _, Some(raxId), _, _) => raxId}).toSet
+
+    if (writeAll && sendQuality) {
+      defaultProject.foreach(headerManager.appendHeader(OpenStackIdentityV3Headers.X_PROJECT_ID, _, 1.0))
+      projectsFromRoles foreach { rolePid =>
```

Comment:
Actually, I kind of don't like how it looks once I started doing it. If you're fine with leaving it as-is for now, then I think that's what I'll do.